label_studio/projects/models.py

```diff
@@ -451,6 +452,47 @@ def _update_tasks_states(
         elif tasks_number_changed and self.overlap_cohort_percentage < 100 and self.maximum_annotations > 1:
             self._rearrange_overlap_cohort()
 
+    def _batch_update_with_retry(self, queryset, batch_size=500, max_retries=3, **update_fields):
+        """
+        Update tasks in batches with retry logic to handle deadlocks.
+
+        Args:
+            queryset: QuerySet of tasks to update
+            batch_size: Number of tasks to update in each batch
+            max_retries: Maximum number of retry attempts for each batch
+            **update_fields: Fields to update (e.g., overlap=1)
+        """
+        task_ids = list(queryset.values_list('id', flat=True))
+        total_tasks = len(task_ids)
+
+        for i in range(0, total_tasks, batch_size):
+            batch_ids = task_ids[i : i + batch_size]
+            retry_count = 0
+
+            while retry_count < max_retries:
+                try:
+                    with transaction.atomic():
+                        Task.objects.filter(id__in=batch_ids).update(**update_fields)
+                    break
+                except OperationalError as e:
+                    if 'deadlock detected' in str(e):
+                        retry_count += 1
+                        if retry_count >= max_retries:
+                            logger.error(
+                                f'Failed to update batch after {max_retries} retries. '
+                                f'Project: {self.id}, Batch: {i}-{i+len(batch_ids)}, Error: {e}'
+                            )
+                            raise
```

Comment:
I think it would be better for readability to move this to after the while loop exits, so that you don't have to compare retry count and max retries twice